src/member.rs

```diff
@@ -0,0 +1,467 @@
+use std::collections::{btree_set::BTreeSet, HashMap};
+use std::net::SocketAddr;
+use std::sync::mpsc::channel;
+use std::sync::{Arc, Mutex};
+use std::thread;
+use std::thread::JoinHandle;
+use std::time::Duration;
+
+pub struct Member {
+    inner: Arc<Mutex<Inner>>,
+}
+
+impl Member {
+    /// Setup a Member to start DKG
+    pub fn new<R: RngCore>(config: Config, rng: &mut R) -> Result<Self, Error> {
+        let node = KeyInfo::new();
+
+        let (node_tx, node_rx) = unbounded::<Event>();
+        let (client_tx, _client_rx) = unbounded();
+
+        let quic_p2p = Builder::new(quic_p2p::EventSenders { node_tx, client_tx })
+            .with_config(config)
+            .build()
+            .map_err(|e| Error::QuicP2P(format!("{:#?}", e)))?;
+
+        let inner = Inner {
+            quic_p2p,
+            id: rng.gen(),
+            group: Default::default(),
+            key_gen: None,
+            our_keys: node,
+            non_responsive: false,
+        };
+        let arc_inner = Arc::new(Mutex::new(inner));
+
+        let _ = setup_quic_p2p_event_loop(&arc_inner, node_rx);
+
+        Ok(Self { inner: arc_inner })
+    }
+
+    /// Connects to every member in the given group
+    pub fn connect_to_group(&mut self, group: HashMap<NodeID, SocketAddr>) {
+        let mut inner_locked = self.inner.lock().unwrap();
+
+        // Required to for testing network churns
+        #[cfg(test)]
+        inner_locked.disconnect_from_all();
+
+        inner_locked.group = group;
+        inner_locked.connect_to_all();
+    }
+
+    /// Initialize DKG with the connected nodes and returns back the proposal
+    pub fn init_dkg(&mut self, threshold: usize) -> Result<Message<NodeID>, Error> {
+        let mut inner_locked = self.inner.lock().unwrap();
+
+        // Extract public_keys of the nodes from the given group for DKG
+        let pub_keys = inner_locked
+            .group
+            .keys()
+            .fold(BTreeSet::new(), |mut set, key| {
+                let _ = set.insert(key.clone());
+                set
+            });
+
+        // Initialize DKG
+        let (key_gen, broadcast_msg) =
+            KeyGen::initialize(&inner_locked.our_keys, threshold, pub_keys)?;
+
+        inner_locked.key_gen = Some(key_gen);
+        Ok(broadcast_msg)
+    }
+
+    /// Broadcast given message to all the connected peers
+    pub fn broadcast(&mut self, msg: Message<NodeID>) -> Result<(), Error> {
+        self.inner.lock().unwrap().broadcast(msg)
+    }
+
+    /// Begins timed phase transition
+    pub fn start_timed_phase_transition(&mut self) -> Result<Vec<Message<NodeID>>, Error> {
+        self.inner.lock().unwrap().start_timed_phase_trasition()
+    }
+
+    /// Fetches our NodeID
+    pub fn id(&self) -> NodeID {
+        self.inner.clone().lock().unwrap().our_keys.node_id()
+    }
+
+    /// Check if our node has received all contributions
+    #[cfg(test)]
+    pub fn all_contribution_received(&self) -> Result<bool, Error> {
```

Comment:
when there is mutiple test only functions, we normally group them into a separate impl brackets, which can then be guarded by only one `cfg(test)`